src/NodeApi/JSValue.cs

```diff
@@ -302,27 +282,1121 @@ public static JSValue CreateDataView(int length, JSValue arrayBuffer, int byteOf
+        if (status == napi_status.napi_invalid_arg)
+        {
+            return null;
+        }
+
+        status.ThrowIfFailed();
+        return GCHandle.FromIntPtr(result).Target;
+    }
+
+    /// <summary>
+    /// Gets the object that was previously attached to this JSValue.
+    /// (Throws an exception if unwrapping failed.)
+    /// </summary>
+    public object Unwrap(string? unwrapType = null)
+    {
+        JSRuntime runtime = GetRuntime(out napi_env env, out napi_value handle);
+        napi_status status = runtime.Unwrap(env, handle, out nint result);
+
+        if (status == napi_status.napi_invalid_arg && unwrapType != null)
+        {
+            throw new JSException(new JSError($"Failed to unwrap object of type '{unwrapType}'"));
+        }
+
+        status.ThrowIfFailed();
+        return GCHandle.FromIntPtr(result).Target!;
+    }
+
+    /// <summary>
+    /// Detaches an object from this JSValue.
+    /// </summary>
+    /// <param name="value">Returns the wrapped object, or null if nothing was wrapped.</param>
+    /// <returns>True if a wrapped object was found and removed, else false.</returns>
+    public bool RemoveWrap(out object? value)
+    {
+        JSRuntime runtime = GetRuntime(out napi_env env, out napi_value handle);
+        napi_status status = runtime.RemoveWrap(env, handle, out nint result);
+
+        // The invalid arg error code is returned if there was nothing to remove.
+        if (status == napi_status.napi_invalid_arg)
+        {
+            value = null;
+            return false;
+        }
+
+        status.ThrowIfFailed();
+        value = GCHandle.FromIntPtr(result).Target;
+        return true;
+    }
+
+    /// <summary>
+    /// Gets the object that is represented as an external value.
+    /// (Throws if the JS value is not an external value.)
+    /// </summary>
+    public unsafe object GetValueExternal()
+    {
+        JSRuntime runtime = GetRuntime(out napi_env env, out napi_value handle);
+        runtime.GetValueExternal(env, handle, out nint result).ThrowIfFailed();
+        return GCHandle.FromIntPtr(result).Target!;
+    }
+
+    /// <summary>
+    /// Gets the object that is represented as an external value, or null if the JS value
+    /// is not an external value.
+    /// </summary>
+    public unsafe object? TryGetValueExternal()
+    {
+        JSRuntime runtime = GetRuntime(out napi_env env, out napi_value handle);
+        napi_status status = runtime.GetValueExternal(env, handle, out nint result);
+
+        // The invalid arg error code is returned if there was no external value.
+        if (status == napi_status.napi_invalid_arg)
+        {
+            return null;
+        }
+
+        status.ThrowIfFailed();
+        return GCHandle.FromIntPtr(result).Target!;
+    }
+
+    public JSReference CreateReference() => new(this);
+
+    public JSReference CreateWeakReference() => new(this, isWeak: true);
```

Comment:
Should these 2 methods be combined into `CreateReference(bool isWeak = false)`?

Reply:
We can add the optional `isWeak` parameter to `CreateReference` but having the separate `CreateWeakReference` is great for the API discoverability.

Reply:
The `JSReference` constructor has a `bool isWeak = false` parameter. I'd prefer to be consistent with that, and an additional `CreateWeakReference()` just seems redundant. If you're concerned about discoverability, maybe the parameter should not be optional.

Reply:
Should we remove these methods from `JSValue`?

Reply:
Probably - they are not any more convenient than the `JSReference` constructor.

Reply:
They are removed.